### PR TITLE
Support `dtype` kwarg in `_foreach_norm`

### LINKED ISSUE #125040: [Feature Request] Support `dtype` arg in `torch._foreach_norm`
`torch.linalg.vector_norm()` and `torch.norm` support the `dtype` argument. Can we get support for it in `torch._foreach_norm`?

cc @crcrpar @mcarilli @janeyx99

### PR DESCRIPTION
**Pull request by crcrpar**
Fixes #125040

cc @awgu

### CODE REVIEW

**Comment by crcrpar** [comment]
marking as draft because this is based on 125527

**Review by janeyx99 — COMMENTED** [review]
Took a look at the changes, they seem reasonable but the meta test failures are real.

Typically I think we've skipped when the original OpInfos are broken, but I have not dug into the problem here so it may be a different concern.

**Comment by crcrpar** [comment]
> Took a look at the changes, they seem reasonable but the meta test failures are real.
> 
> Typically I think we've skipped when the original OpInfos are broken, but I have not dug into the problem here so it may be a different concern.

yes, they are real failures but not that opinfos are broken. The cause is that `dtype` parameterization which is kind of hidden in test_foreach's parity check

**Comment by janeyx99** [comment]
> yes, they are real failures but not that opinfos are broken. The cause is that `dtype` parameterization which is kind of hidden in test_foreach's parity check

Could you elaborate what you mean by this? Or will you be pushing a fix for this?

**Comment by crcrpar** [comment]
> > yes, they are real failures but not that opinfos are broken. The cause is that `dtype` parameterization which is kind of hidden in test_foreach's parity check
> 
> 
> 
> Could you elaborate what you mean by this? Or will you be pushing a fix for this?
> 
> 

pushed a fix🙂

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -9373,12 +9373,10 @@ def __call__(self, opinfo, device, dtype, requires_grad, **kwargs):
         _foreach_inputs_kwargs = {k: kwargs.pop(k, v) for k, v in _foreach_inputs_default_kwargs.items()}
         _foreach_inputs_kwargs["requires_grad"] = requires_grad
 
-        from torch.testing._internal.common_dtype import get_all_dtypes
-
         for num_tensors, ord, out_dtype in product(
             num_input_tensors,
             (0, 1, 2, -1, -2, float('inf'), float('-inf')),
-            [None] + get_all_dtypes(),
+            (None,) + (torch.complex128,) if dtype in complex_types() else (torch.float64,),
```
Comment (janeyx99) [inline torch/testing/_internal/common_methods_invocations.py:9379]: what coverage does this remove?

Comment (crcrpar) [reply torch/testing/_internal/common_methods_invocations.py:9379]: integers, complex64, float32, float16, and bfloat16. 
integers wouldn't be supported anyway, i.e., test_foreach's test_parity checks the error messages are identical.
some lower prec's could cause no direct cast or something

Comment (janeyx99) [reply torch/testing/_internal/common_methods_invocations.py:9379]: I'd rather still test more + just have expected skips vs silently testing less

Comment (crcrpar) [reply torch/testing/_internal/common_methods_invocations.py:9379]: I found it not really straightforward to give skips/xfail to some of the sample inputs generated.
Would you happen to have some reference?

Comment (janeyx99) [reply torch/testing/_internal/common_methods_invocations.py:9379]: https://github.com/pytorch/pytorch/blob/95b2766864a02cdb6bc49b97a7f7e859b7167e70/torch/testing/_internal/common_methods_invocations.py#L9481-L9507

something like this but for norm

Comment (crcrpar) [reply torch/testing/_internal/common_methods_invocations.py:9379]: I'm sure my previous response is misleading. The dtypes with if-else here doesn’t eliminate any of the existing test cases as it’s aimed to test the new kwarg. Inappropriate dtype to the kwarg just leads to an error, and unfortunately for meta test, all dtypes (meaning parameterization of inputs) end up being decorated as xfail if the dtype kwarg parameterization covers all the possible ones

Comment (janeyx99) [reply torch/testing/_internal/common_methods_invocations.py:9379]: Ah this parametrizes out_dtype, which is new.

**Review by janeyx99 — APPROVED** [review]

**Comment by janeyx99** [comment]
@pytorchbot merge -i

ROCm dist failures seem not related.

**Comment by crcrpar** [comment]
is this still being checked...? The checks look okay

**Comment by janeyx99** [comment]
Ah this was merged but just not closed on Wednesday around 4pm. Closing as this commit has been merged.

**Pull request closed without merging**